### PR TITLE
Proposal: Mirroring Support for Private Registries

### PR DESCRIPTION
**Pull request by jlhawn**
Docker-DCO-1.1-Signed-off-by: Josh Hawn josh.hawn@docker.com (github: jlhawn)

### CODE REVIEW

**Comment by stevenschlansker** [comment]
@SvenDowideit any chance of this getting merged soon?

**Comment by tarnfeld** [comment]
> (eg: people expected significantly better performance thanks to mirroring - which it failed to deliver in a number of cases, given perf issues are rather engine-side)

I disagree completely here, though sure there may be some performance issues on the Docker side, for those that are using docker across multiple regions pay a pretty big penalty in latency, throughput and cost for shipping images across continents. To avoid an incredibly confusing setup with multiple registries and having users push their image to multiple places and keeping it all in sync, it's easier to have a single master registry (at least without a lot of extra tooling) and that means systems running in multiple sites or regions pay the price of slower image pulls.

If the V2 protocol is coming along nicely, this can certainly be a consideration in that design, but why do we need to hold this change up? It's a pretty trivial change and this code is likely going to be removed or changed for V2 anyway?

**Comment by stevenschlansker** [comment]
@dmp42 Yes, there's a ton of speed problems.  This is just addressing one of them and believe me I'm fully aware that `docker push` and `docker pull` will still be glacially slow.  But at least now it'll be glacier-with-global-warming slow instead of fully frozen :P

This is about as trivial of a change as it gets - the `registry-mirror` option before only worked for mirrors of the public Docker registry, but now it works with private mirrors as well.  The change makes no new policy decisions about authentication, private content, or any other 'sensitive' items.  I even added documentation of the (already mostly existing) behavior which was previously undocumented!

I'm also not sure why you want a proposal at this point - this PR has been open for 2 months already and all a proposal will be is "Propose to do what this patch over here does."  At the very least hearing that feedback in November would be a lot less frustrating than hearing it now.

I'm sure the V2 registry will be great but do understand that we've already been waiting most of a year for any improvements to speed and an answer of 'just wait for the Next Great Thing' really leaves your users out in the cold :(

**Comment by shin-** [comment]
Sorry for butting in, can someone give me a quick explanation of what this PR provides that [registry mirroring options](https://github.com/docker/docker-registry/#mirroring-options) didn't already? I'm a bit out of the loop I'm afraid.

**Comment by stevenschlansker** [comment]
@shin- They work together.  For example, the setup I want is:

`central.company.com` is our "source of truth" registry.
`mirror-us.company.com`
`mirror-eu.company.com` are mirrors of `central` via the mirroring options you indicate.

Then, the Docker clients are configured e.g.
`--registry-mirror=mirror-us.company.com --registry-mirror=central.company.com` for US instances, so they first try the local mirror but fall back to `central` if it is down or unavailable.

**Comment by jlhawn** [comment]
@shin- said:

> Sorry for butting in, can someone give me a quick explanation of what this PR provides that registry mirroring options didn't already? I'm a bit out of the loop I'm afraid.

Currently, the mirroring options only allow you to mirror official repositories from `index.docker.io`. This PR allows mirroring outside of this condition.

**Comment by shin-** [comment]
@jlhawn FWIW You should be able to set mirroring.source to a private registry URL. But I do understand @stevenschlansker 's use case.
Thanks - as far as I can tell, this is not exclusive to or tied into registry v1, and the change is relatively trivial, so I see no reason to refuse it on those grounds.

**Comment by jlhawn** [comment]
@shin- said:

> FWIW You should be able to set mirroring.source to a private registry URL

Yup, but the configured mirrors have been getting ignored when pulling a non-docker-hub image.

**Comment by dmp42** [comment]
@tarnfeld 

>> (eg: people expected significantly better performance thanks to mirroring - which it failed to deliver in a number of cases, given perf issues are rather engine-side)

> I disagree completely here, though sure there may be some performance issues on the Docker side, 

Please see https://github.com/docker/docker-registry/issues/785

@jlhawn 

> Yup, but the configured mirrors have been getting ignored when pulling a non-docker-hub image.

For a good reason. Private content / authnz do not work well / is unspecified with mirroring as it is designed right now.
Also raise considerations about the integrity of the namespace.

@shin-
 
* there is no documentation (or support, see below) for private registries mirroring on docker-registry
* it's largely untested (see Sven comment: "I'm assuming that @stevenschlansker has tested this PR")
* behavior is unspecified with regard to authentication, private / public content, mirror chaining...
* the security angle is unevaluated as far as I can tell (what happens if I bind to index.docker.io and prompt the user for their credentials? can I as a mirror gain access to private content?)
* impact on the global namespace is unevaluated
* this is tied into v1, and if this is merged, will have to be supported, no matter how good or bad its design is (meaning: it should be good)

I'm *not opposed* to this feature in itself - it just needs to be worked on further.

Fixing problems that people have is definitely a good thing - but quality is important too.

Hope that clarifies.

**Comment by tarnfeld** [comment]
> (eg: people expected significantly better performance thanks to mirroring - which it failed to deliver in a number of cases, given perf issues are rather engine-side)

> I disagree completely here, though sure there may be some performance issues on the Docker side,

> Please see docker/docker-registry#785

Well in the situation I was testing, I verified the client was hitting the primary registry, so we're talking about different issues. Also, the network you're operating in and across will hugely affect the behaviour, especially if the link to the primary registry is so drastically slow the speed difference is clear.

**Comment by tarnfeld** [comment]
That said, I do agree with the statements about authentication.

**Comment by dmp42** [comment]
> Well in the situation I was testing, I verified the client was hitting the primary registry, so we're talking about different issues. Also, the network you're operating in and across will hugely affect the behaviour, especially if the link to the primary registry is so drastically slow the speed difference is clear.

@tarnfeld that person was hitting their mirror as well, and everything was configured like it should - still pulling from their mirror was not any faster than from the hub itself, for the reason I mentioned (lies in docker itself, see below).

The bottom-line being:

 * people with terrible links, like Sven, (X<1MB) *will enjoy* speed benefits from mirroring (from X up to 1MB)
 * people with a link that is able to deliver anything above 1MB/sec (and this is not much) will likely not enjoy any speed improvement

The good news is that speed improvements are being worked on, actively: https://github.com/docker/docker/pull/9720

About bandwidth, certainly, you can definitely save on that, there is no argument about that - I'm interested in knowing more about people usage on that front - how much bandwidth they push a month through their private registry? how much does that cost?

**Comment by stevenschlansker** [comment]
@dmp42 The behavior today is to *only* use mirrors instead of talking to the public registry, which is the only case that login credentials are relevant, and that behavior exists in the release today.  This change only allows mirroring of *private* registries, which don't interact with `index.docker.io` as you worry about.

That also brings up some scary implications, namely:
* docker will send login credentials to servers which are not correctly signed as `index.docker.io`?
* docker sends credentials in cleartext rather than some sort of challenge/response?!

Also, the administrator has to explicitly configure these mirrors on the docker daemon.  IMO it is therefore up to the administrator to ensure that their mirrors are protected well otherwise arbitrary content could be served anyway, aside from the worry about credentials.

As for the `v1` vs `v2` concerns, I'm happy to have this feature change / disappear / anything after the v2 registry is stable and usable and covers our use cases.

**Comment by tarnfeld** [comment]
> @tarnfeld that person was hitting their mirror as well, and everything was configured like it should - still pulling from their mirror was not any faster than from the hub itself, for the reason I mentioned (lies in docker itself, see below).

Sorry perhaps I didn't explain my point well enough. The testing I did resulted in the mirror *never* being hit and I verified the code path with verbose logging, the mirroring code never gets hit in the client hence why this fix is being proposed. The client always went to the primary registry directly regardless of the mirror configuration, because the code does not allow private registries to go via mirrors.

**Comment by shin-** [comment]
There is a lot of confusion here regarding what we call mirrors in the docker daemon, and what we call mirrors in the docker-registry code, which makes the conversation all the more difficult to follow.

**Comment by SvenDowideit** [comment]
as an anecdote :)

before the --registry-mirror code was released working, hub downloads of the golang image used up about 80% of my bandwidth quota for the preceding month or 2 (whenever we started using it for boot2docker - now, it is not quite irrelevant.

each fresh test of a boot2docker vm would re-download the docker project build images, which would take in the order of an hour, whereas now, its close enough to instant that i don't have time to make another coffee.

so I imagine for a distributed organisation using private registries for their internal workflows, the problem is just as big a deal.

**Comment by bobrik** [comment]
Imagine that you deploy 800mb app (2 layers) 20-40 times a day to 150 machines daily. You deploy it from private registry and you also have mirror set up because docker hub is giving you images at 14kbps.

Right now workflow is:

1. Push image to local registry
2. Pull image from local registry on every node directly

With this change:

1. Push image to local registry
2. Pull image from local mirror, figure out that it's not there (images deployed once, why would it be there?), start mirroring in 150 parallel forks.
3. If that worked, pull image from mirror that is not a bit faster than local registry because it shares the same distributed storage backend.

Why do that? I don't see a reason to break existing behavior, since we cannot disable mirroring for local registries.

If you really need to mirror private registries — go ahead and set up caching nginx in front of your private registry, it is just highly cacheable http service after all. You have a solution. I don't see which solution do I have except "avoid docker upgrades before v2 registry is out".

**Comment by bobrik** [comment]
Your "private mirror" could be replaced with nginx that only does caching (with locking) for `layer` files. Just point private registry domain to that nginx in every location. I fully understand you goals and desire to do that with registry itself.

My concern is that now we are breaking my existing setup without giving me an ability to fix it. I have 1gb link to my mirror and 1gb link to my registry.

`[host] - 1gb -> [local registry]` changes to `[host] - 1gb -> [mirror] - 1gb -> [local registry]`.

Just let me have the old way without an extra hop. I'd love to see mirroring for private registries, but with the way to disable it for local ones.

**Comment by tarnfeld** [comment]
@bobrik This is an optional configuration. A mirror is not _required_ to be able to use a private registry, nor will it be with this change.

> I fully understand you goals and desire to do that with registry itself.

This change has nothing to do with the actual registry, this is a configuration option in the docker client which allows you to tell it to point to the `[mirror]`. Without this, your example `[host] - 1gb -> [mirror] - 1gb -> [local registry]` is impossible.

Is your concern that you would like to use your registry as a mirror for index.docker.io BUT not for your own private images?

**Comment by bobrik** [comment]
> Is your concern that you would like to use your registry as a mirror for index.docker.io BUT not for your own private images?

Exactly. I want to use mirroring registry for index.docker.io and all private registries, except my own registry that cannot become faster with mirroring.

**Comment by tarnfeld** [comment]
I see, that makes sense. Intriguing that you're using multiple private registries but that's a legit issue. @stevenschlansker any thoughts?

**Comment by stevenschlansker** [comment]
@bobrik @tarnfeld If your mirror has a mirror of public registry content but not your private registry, then when it fails to serve the image Docker will fall back to the 'real' private registry that you want it to read from.  So as long as your 'mirror of index.docker.io' registry does not mirror your actual content, the only regression here is that it must check once that the image is not there before doing the same pull as it always did.
Do you agree?

**Comment by bobrik** [comment]
I agree. How do I do that?

File: docs/sources/reference/commandline/cli.md [hunk 1]
```diff
@@ -333,6 +333,14 @@ To set the DNS server for all Docker containers, use
 To set the DNS search domain for all Docker containers, use
 `docker -d --dns-search example.com`.
 
+### Daemon registry-mirror option
+
+To save on network bandwidth, it is useful to cache images close to the Docker
+instances using them. When an image is pulled from its upstream registry, if
+one or more `--registry-mirror=http://<my-docker-mirror-host>` options are
+specified, the given mirrors are checked in order to see if they have a cached
```
Comment (fredlf) [inline docs/sources/reference/commandline/cli.md]: Where "in order" is left to right? Alphabetical?

**Comment by fredlf** [comment]
A copy-edit and a minor question. Once those nits are picked, docs LGTM.

**Comment by stevenschlansker** [comment]
It looks like the merge of https://github.com/docker/docker/pull/8456 causes this entire patch to end up as a merge conflict.

It looks like that PR changes how mirrors are handled, without mentioning that fact at all in the description.  I did a brief read-through of #8456 but can't see how to resolve this patch against that one.  And I'm not terribly thrilled about rebasing this patch *again*.  Maybe someone who knows Docker better than I do will care enough to rebase.

**Comment by crosbymichael** [comment]
@stevenschlansker 

I'm a little confused right now, I don't see any code in this PR right now.  Did you remove those commits or what's up?

I'm sure someone else or one of the maintainers can rebase the PR for you.

**Comment by jlhawn** [comment]
@crosbymichael Yes, I removed the code commit for now because https://github.com/docker/docker/pull/8456 made it very different and now explicitly sets mirroring for the official "indexInfo" config. We'll have to rethink how to specify mirrors for private registries - probably a new CLI option on the daemon.

**Comment by jamtur01** [comment]
Docs LGTM

**Comment by icecrime** [comment]
Closing this as it is now tracked on docker/distribution (I guess docker/distribution#19 is a good entrypoint).

**Pull request closed without merging**